File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    static boolean[] visited;
    static int[][] graph;
    
    static int bfs(int n, int c){
        visited = new boolean[n+1];
        int cnt = 1;
        Queue<Integer> q = new LinkedList<>();
        visited[c] = true;
        q.offer(c);
        while(!q.isEmpty()){
            int t = q.poll();
            for(int i=0; i<n+1;i++){
                if(graph[t][i] == 1 && !visited[i]){
                    visited[i] = true;
                    q.offer(i);
                    cnt++;
                }
            }
        }
        return (int)Math.abs(2*cnt - n);
    }
    public int solution(int n, int[][] wires) {
        int ans = (int)1e9;
        graph = new int[n+1][n+1];
        
        for(int[] nums: wires){
            graph[nums[0]][nums[1]] = 1;
            graph[nums[1]][nums[0]] = 1;
        }
        
        for(int[] nums: wires){
            graph[nums[0]][nums[1]] = 0;
            graph[nums[1]][nums[0]] = 0;
            
            ans = Math.min(ans,bfs(n,nums[0]));
            
            graph[nums[0]][nums[1]] = 1;
            graph[nums[1]][nums[0]] = 1;
        }
        
        return ans;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict, deque    
def solution(n, wires):
    graph = [[] for _ in range(n+1)]
    
    for s, e in wires:
        graph[s].append(e)
        graph[e].append(s)
    
    def bfs(start, cut):
        visited = set()
        q = deque([start])
        visited.add(start)
        while q:
            node = q.popleft()
            for nxt in graph[node]:
                if nxt in cut and node in cut:
                    continue
                if nxt not in visited:
                    visited.add(nxt)
                    q.append(nxt)
        return len(visited)
    
    cur = n
    for s, e in wires:
        cnt = bfs(s, (s,e))
        dif = abs(n- 2*cnt)
        cur = min(cur,dif)
    return cur